opa-audit-operator/src/charm.py

```diff
@@ -0,0 +1,177 @@
+
+    @property
+    def name(self):
+        return self._name
+
+
+class OPAAuditCharm(CharmBase):
+    """
+    A Juju Charm for OPA
+    """
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.stop, self._on_stop)
+        self.framework.observe(self.on.install, self._on_install)
+        self._stored.set_default(things=[])
+
+    def _on_config_changed(self, _):
+        """
+        Set a new Juju pod specification
+        """
+        self._configure_pod()
+
+    def _on_stop(self, _):
+        """
+        Mark unit is inactive
+        """
+        self.unit.status = MaintenanceStatus("Pod is terminating.")
+        logger.info("Pod is terminating.")
+
+    def _on_install(self, event):
+        logger.info("Congratulations, the charm was properly installed!")
+
+    def _on_update_status(self, event):
+        logger.info("Status updated")
+
+    def _build_pod_spec(self):
+        """
+        Construct a Juju pod specification for OPA
+        """
+        logger.debug("Building Pod Spec")
+        crds = []
+        try:
+            crds = [
+                yaml.load(Path(f).read_text())
+                for f in [
+                    "files/configs.config.gatekeeper.sh.yaml",
+                    "files/constrainttemplates.templates.gatekeeper.sh.yaml",
+                    "files/constraintpodstatuses.status.gatekeeper.sh.yaml",
+                    "files/constrainttemplatepodstatuses.status.gatekeeper.sh.yaml",
+                    # "files/sync.yaml",
+                ]
+            ]
+        except yaml.YAMLError as exc:
+            print("Error in configuration file:", exc)
+
+        crd_objects = [
+            CustomResourceDefintion(crd["metadata"]["name"], crd["spec"])
+            for crd in crds
+        ]
+
+        config = self.model.config
+        spec_template = {}
+        with open("files/pod-spec.yaml.jinja2") as fh:
+            spec_template = Template(fh.read())
+
+        import pdb; pdb.set_trace()
+        template_args = {
+            "crds": crd_objects,
+            "image_path": config["imagePath"],
+            "imagePullPolicy": config["imagePullPolicy"],
+            "app_name": self.app.name,
+            "audit_cli_args": self._audit_cli_args(),
+            "namespace": os.environ["JUJU_MODEL_NAME"],
+        }
+
+        spec = yaml.load(spec_template.render(**template_args))
+
+        print(f"Pod spec: {spec}")
+        return spec
+
+    def _audit_cli_args(self):
+        """
+        Construct command line arguments for OPA Audit
+        """
+
+        args = [
+            "--operation=audit",
+            "--operation=status",
+            "--logtostderr",
+        ]
+
+        return args
+
+    def _opa_config(self):
```

Comment:
Where is _opa_config called? Is it dead code?

Reply:
Please remove this.

Reply:
removed